### DISCUSSION TITLE
Pattern overlaying on clothing

### DISCUSSION THREAD

**Discussion by DOEHOONLEE** [ImageMagick/ImageMagick#4177] [Help]
Hi!

I have been looking into [this post](https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=33251) lately, but I am still a little bit confused with the purpose of displace and lighting images introduced in it.

One necessary step for creating displace and lighting images were this.

```bash
color=`convert clothing.png mask.png -alpha off -compose copy_opacity -composite \
-scale 1x1! -alpha off -type grayscale -format "%[pixel:u.p{0,0}]" info: | tr -cs "0-9*\n" " "`
diffl=`convert xc: -format "%[fx:(40-100*($color/255))]" info:`
diffd=`convert xc: -format "%[fx:(50-100*($color/255))]" info:`
echo "color=$color; diffl=$diffl; diffd=$diffd"
color= 227 ; diffl=-49.0196; diffd=-39.0196
```

The putting pattern on clothing still worked without all the mask, displace, lighting images, but it does look a little different.

Can someone explain what the calculated values are for `diffl` and `diffd`?

**Comment by snibgo** [comment]
> I have been looking into this post lately,

What post?

**Reply by DOEHOONLEE** [reply]
ohh I am so sorry I forgot to include the link. I just edited it.

**Reply by DOEHOONLEE** [reply]
Someone on StackOverflow said that
`xc: used to mean "X Constant Image" but now is just a shorthand for canvas:.`

but I still cannot figure out what either `xc:` or `canvas:` can do :(

**Reply by fmw42** [reply]
It is used with -size WxH to create a constant color image. For example,

`-size 100x100 xc:red`

creates a 100x100 red image.

**Reply by DOEHOONLEE** [reply]
What do you mean by constant color? like single colored image?

**Comment by fmw42** [comment]
I mean the whole image is the same unique color - there is no variation any where in the image in that color.

**Reply by DOEHOONLEE** [reply]
oh, like a finger print to each person?

**Reply by DOEHOONLEE** [reply]
but then what are these for?

```bash
-scale 1x1! -alpha off -type grayscale -format "%[pixel:u.p{0,0}]" info: | tr -cs "0-9*\n" " "`
diffl=`convert xc: -format "%[fx:(40-100*($color/255))]" info:`
diffd=`convert xc: -format "%[fx:(50-100*($color/255))]" info:`
```

like `%[pixel:u.p{0,0}]`? or `%[fx:]` ?? ending with `info:` ?? I am so confused..

I really wish there were detailed documentations :(

**Reply by fmw42** [reply]
`-format "%[pixel:u.p{0,0}]"` returns the color of the first pixel in the image that is being worked on. 

`-scale 1x1! -alpha off -type grayscale -format "%[pixel:u.p{0,0}]" info:`

This averages the image to 1x1 pixel, converts to grayscale and then returns the color of that average pixel.

In the diff lines, you have xc: with no color. That is a placeholder image for doing fx: computations. So 

`"%[fx:(40-100*($color/255))]"`

is taking the color in the variable $color, dividing by 255, multiply by 100 to convert to percent and subtracting from 40.  The %[fx: ... ] is the command to do computations.

**Reply by fmw42** [reply]
I do not know what you mean or asking by

"oh, like a finger print to each person?"

-size 100x100 xc:red  

creates a 100x100 pixel red image

**Reply by DOEHOONLEE** [reply]
So what the above codes are doing is that
(I am still no 100% what they each mean, but these are how I understood after looking through all the pages you provided and your answer here)

(1) `-alpha off` is done every time to ensure that no alpha channel is enable.

(2) `-compose copy_opacity -composite`
takes the grayscaled image's opacity and applies it onto the mask image.

(3) `-scale 1x1!`
ignores the width/height ratio and resizes the image into 1px by 1px.

(4) `-alpha off` again to make sure that no alpha channel is editable.

(5) `-type graysacle` ignores and overrides so that the encoder sees the image with gray scale present only.

(6) `-format "%[pixel:u.p{0,0}]"` takes the the first pixel of the first image and returns the value in percentage.

Am I getting it right?

So up to (6) returns a grayscale value in percentage, for example, like this: gray(89.0536%)%

THEN, `tr -cs "0-9*\n" " "` takes the above value, replaces everything to " " except digits => 89 0536

and the value we got is stored in the variable `$color` (= 89 0536).

**In short**, these steps "applies the same transparency in the grayscale image to the mask then resize the image into 1pxx1px, then take the first pixel of it, convert it to grayscale, then return as grayscale percentage??"

Am I getting it right?

So all these steps are to get the average grayscale value in percentage?


Plus, `convert xc: -format "%[fx:(40-100*($color/255))]" info:` I get what this "does," but I am not sure what it is for..?

Why does one being subtracted from 40 and the other from 50? Is this just some formula to create light and displace images??

**Reply by fmw42** [reply]
1 ) -alpha off is done when one wants to disable the alpha channel. This is needed, for example, when using -compose copy_opacity to add an alpha channel from a second image to avoid conflict with the alpha from the first image. The second -alpha off is used so that the returned color does not have an alpha value , that is, it has only rgb or in this case gray values.

6 ) -format "%[pixel:u.p{0,0}]"  returns the color of the first pixel at (0,0) of the first image, which in this case is the only image left in the command line. Whether it is percent or not depends upon imagemagick. It could also be a color name. But in this case since the image is grayscale, it returns a gray value.

I have no idea what "%[fx:(40-100*($color/255))]" is computing without seeing all of the code to which you refer and to what they are trying to do. How are diffl and diffd used. Typically the former is to adjust the shading of the image using the grayscale image to modulate intensity via hard light composite to create the appearance of folds (brighter and darker regions near where it is displace)  and the second is create the displacement map.  Typically displacement map needs to have gray level = 50% where one does not want it to distort. Darker will distort in one direction and brighter in the opposite direction. So one shifts the brightness of the grayscale image so that it averages 50%.  The adjustment of the lighting apparently is desired by the author to be centered at 40%.  The color terms are the corrections. At least that is how I use them. These two equations appear to compute the difference between 40% and the image gray level in percent for the first and the difference between 50% and the image gray level in percent for the second.

**Reply by DOEHOONLEE** [reply]
In the [reference](https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=33251) I included in my original post,

```bash
color=`convert clothing.png mask.png -alpha off -compose copy_opacity -composite \
-scale 1x1! -alpha off -type grayscale -format "%[pixel:u.p{0,0}]" info: | tr -cs "0-9*\n" " "`
diffl=`convert xc: -format "%[fx:(40-100*($color/255))]" info:`
diffd=`convert xc: -format "%[fx:(50-100*($color/255))]" info:`
echo "color=$color; diffl=$diffl; diffd=$diffd"
color= 227 ; diffl=-49.0196; diffd=-39.0196
```

these were required in order to create lighting and displace images below.

![lighting](https://user-images.githubusercontent.com/20305442/133026034-468e7e56-a3d2-481c-8787-002698093ede.png)
![displace](https://user-images.githubusercontent.com/20305442/133026038-c29ce3e5-cc82-4d24-a885-124ffd664083.png)

Perhaps, they are the way to create lighting and displace images?

**Reply by fmw42** [reply]
That code does not show how diffl and diffd are used, nor how the displacement is created. I need to see the full code from start to finish.

**Reply by DOEHOONLEE** [reply]
ohh I see. Let me include the steps in the post with explanation.

![Screen Shot 2021-09-13 at 2 26 09 PM](https://user-images.githubusercontent.com/20305442/133028501-c607cea9-a3a9-4ab1-9e91-15d9e6c8372f.png)

So the author in this post did the following steps:

1. create a gray scale image
2. create a mask image
3. get color value that I asked earlier
4 & 5. create lighting and displace images using the color value got from 3.
6. create an overlay image using the above.

**Reply by fmw42** [reply]
Please identify the post from which you got this code. What is the link to it?

**Reply by DOEHOONLEE** [reply]
![Screen Shot 2021-09-14 at 2 55 02 PM](https://user-images.githubusercontent.com/20305442/133202641-59944130-48d6-4429-a253-e12859f663da.png)

I am sorry, I edited after you went through the first message. Let me attach here again. Sorry for the confusion.

https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=33251

**Reply by fmw42** [reply]
Yes, that post answer was an old one that I posted. So all the code you are referring to was my code. Are you doing this for personal not for profit use? Or are you working on an app for yourself or a client? The code is an excerpt from one of my ImageMagick scripts. You might want to try using it rather than worry about the details of the code. See my script, outfit, at http://www.fmwconcepts.com/imagemagick. If for profit use, you will have to license it from me, but you can try it for free without a license. Once you start using it for profit, it will need to be licensed.

**Reply by DOEHOONLEE** [reply]
This may be used for profit in the future. Thanks for notifying me. Which part of the post is yours that are licensed? I want to be careful and not violate anything.

**Reply by fmw42** [reply]
Code on this post is not licensed. But if you use my script from my web site, then it  would be licensed.

**Reply by DOEHOONLEE** [reply]
Oh, I see I see. http://www.fmwconcepts.com/imagemagick This website.

I will take a look at it and if I find anything that I would like to adapt in my project, I will make sure to ask for your permission beforehand. Thank you for introducing such a informative website. Great work, by the way!!

**Reply by fmw42** [reply]
See the scripts: outfit, tshirt, cylinderize. They may be of interest if you want an out of the box solution and do not have to develop your own.

**Reply by DOEHOONLEE** [reply]
Amazing resources/ideas!! Thanks a lot for your support and suggestion!

**Comment by fmw42** [comment]
Please review:

https://imagemagick.org/script/command-line-processing.php
https://imagemagick.org/script/fx.php
https://imagemagick.org/Usage/transform/#fx_escapes
https://legacy.imagemagick.org/Usage/basics/
https://imagemagick.org/script/escape.php

**Reply by DOEHOONLEE** [reply]
Thanks!!
I'd only found [this page](https://imagemagick.org/script/escape.php) :(

Thanks again for your help!!
